File: Sean的leetcode/Leetcode/Palindrome Pairs.cpp
//Use hashmap
//Split each word to two part and check whether the palindrome exists
//l    r    l'
//r'   l    r
//if l is empty: it could be l' l  r
//Time: O(n * k ^ 2)
class Solution {
public:
    vector<vector<int>> palindromePairs(vector<string>& words) {
        unordered_map<string, int> table;
        for (int i = 0; i < words.size(); i++) {
            string tmp = words[i];
            reverse(tmp.begin(), tmp.end());
            table[tmp] = i;
        }
        vector<vector<int>> re;
        for (int i = 0; i < words.size(); i++) {
            for (int j = 0; j < words[i].size(); j++){
                string l = words[i].substr(0, j);
                string r = words[i].substr(j);
                if(table.count(l) && table[l] != i && isPalindrome(r)){
                    re.push_back({i, table[l]});
                    if(l.empty()){
                        re.push_back({table[l], i});
                    }
                }
                if(table.count(r) && table[r] != i && isPalindrome(l)){
                    re.push_back({table[r], i});
                }
            }
        }
        return re;
        
    }
    bool isPalindrome(string& s){
        int n = s.size();
        if (n <= 1) return true;
        int l = 0, r = n-1;
        while (l < r) {
            if (s[l] != s[r]) return false;
            l++;
            r--;
        }
        return true;
    }
};

File: MyLeetCode/python/Palindrome Pairs.py
###
# HashTable
# Time Complexity: O(m * n ^ 2) where m is the length of the list and the n is the length of the word.
# Space Complexity: O(m)
###
class Solution(object):
    def palindromePairs(self, words):
        """
        :type words: List[str]
        :rtype: List[List[int]]
        """
        d = {word: i for i, word in enumerate(words)}
        re = []
        for i, w in enumerate(words):
            for j in xrange(len(w)+1):
                str1 = w[:j]
                str2 = w[j:]
                if self.is_palindrome(str1):
                    if str2[::-1] in d and d[str2[::-1]] != i:
                        re.append([d[str2[::-1]], i])
                if str2 and self.is_palindrome(str2):
                    if str1[::-1] in d and d[str1[::-1]] != i:
                        re.append([i, d[str1[::-1]]])
        return re
    def is_palindrome(self, word):
        i = 0
        j = len(word) - 1
        while i <= j:
            if word[i] != word[j]:
                return False
            i += 1
            j -= 1
        return True
                                        
                            
                    